ipld/retriever.go

```diff
@@ -154,9 +156,11 @@ func (rs *retrieverSession) retrieve(ctx context.Context, q *quadrant) (*rsmt2d.
 	rs.request(ctx, q)
 
 	// try repair
+	// TODO: Avoid reimporting of the square which can potentially remove the requirement for the lock
+	rs.squareLk.Lock()
```

Comment:
I'm a bit confused.
1. `rs.request` call in line 156 is synchronous (there is a `WaitGroup` inside). 
2. `put` func passed to `GetShares` is invoked by workerpool, but `GetShares` is also synchronous (there's `WaitGroup` inside).
3. therefore, before reaching this line, all "puts" should already be finished - or I'm missing something?
4. `retrieve` function is also synchronous, and it's only called [in a loop](https://github.com/celestiaorg/celestia-node/pull/682/files#diff-32233bd82783aec17000ced7c82d67d3626a9f344eccde8f6b1c67f7273bba7bR58) - so only one instance of `retrieve` is running at any given point of time.

All above means, that lock is not needed at all. Correct me if I'm wrong.

Reply:
Believe me or not, I thought the same, but the race detector told me I was wrong and this locking fixed the thing. 

The race condition happens specifically when we reimport the square after the second repair. I also placed a todo that after rsmt2d update the race might disappear. To know exactly, we need to understand race detector better. The rsmt2d update is about importing a square once and repair with a method on the square. I can give more details about importing or else in rsmt2d if you are interested.

Reply:
Happy that someone is finally reviewing the concurrency code😌

Reply:
@tzdybal was able to reproduce the race locally to prove that lock is required, so I am resolving the issue